#include <boost/asio.hpp>
#include <iostream>
#include <vector>
#include <string>
#include <cstring>
#include <thread>
#include <mutex>
#include <future>      // å¼•å…¥ future/promise
#include <map>         // å¼•å…¥ map
#include <atomic>      // å¼•å…¥ atomic
#include <arpa/inet.h> // for htonl, ntohl, htons, ntohs
#include <memory>

// å‡è®¾è¿™äº›å¤´æ–‡ä»¶å­˜åœ¨äºæ‚¨çš„ç¯å¢ƒä¸­
#include "protocol.pb.h"
#include "public.h"

using boost::asio::ip::tcp;

// =====================================================================
// å…¨å±€ I/O ç»“æ„ (ç”¨äºå¼‚æ­¥é€šä¿¡)
// =====================================================================

// å½“å‰ç™»å½•ç”¨æˆ·ID
int Currentid_ = -1;

// ä¿æŠ¤åŒæ­¥å†™å…¥æ“ä½œï¼Œé˜²æ­¢å¤šçº¿ç¨‹åŒæ—¶è°ƒç”¨ boost::asio::write
std::mutex g_write_mutex;

// è¯·æ±‚ ID è®¡æ•°å™¨ï¼Œç”¨äºåŒ¹é… Request å’Œ Response
std::atomic<uint32_t> g_request_counter{0};

// ç”¨äºåŒæ­¥ç­‰å¾…å¼‚æ­¥å“åº”çš„ Future/Promise ç»“æ„
struct ResponseWaiter
{
    uint32_t request_id;
    uint16_t expected_msgid;
    // promise ç”¨äºåœ¨å¼‚æ­¥çº¿ç¨‹ä¸­è®¾ç½®ç»“æœï¼Œfuture ç”¨äºåœ¨åŒæ­¥çº¿ç¨‹ä¸­é˜»å¡è·å–ç»“æœ
    std::promise<std::vector<char>> promise;
};

// å­˜å‚¨æ‰€æœ‰æ­£åœ¨ç­‰å¾…çš„å“åº”ï¼šKey=è¯·æ±‚ID
// æ³¨æ„ï¼šç”Ÿäº§ç¯å¢ƒä¸­ï¼ŒRequest ID åº”è¯¥åœ¨åè®®ä¸­ä¼ è¾“ã€‚
// é‰´äºæ‚¨çš„åè®®å›ºå®šï¼Œæˆ‘ä»¬ä½¿ç”¨ä¸€ä¸ªç®€åŒ–æœºåˆ¶ï¼šå‡è®¾åªæœ‰ä¸€ä¸ªåŒæ­¥è¯·æ±‚åœ¨è¿›è¡Œã€‚
// ä¸ºäº†æ¼”ç¤ºå¼‚æ­¥å’Œ futureï¼Œæˆ‘ä»¬å°†ä½¿ç”¨ mapï¼Œä½†ä¾èµ–äºä¸šåŠ¡å±‚ç¡®ä¿ä¸²è¡Œè¯·æ±‚ã€‚
std::mutex g_waiters_mutex;
// ä½¿ç”¨ shared_ptr ç¡®ä¿ promise åœ¨å¼‚æ­¥çº¿ç¨‹ä¸­å¯ä»¥å®‰å…¨è®¿é—®
std::map<uint32_t, std::shared_ptr<ResponseWaiter>> g_response_waiters;

// =====================================================================
// åè®®å·¥å…·å‡½æ•° (ä¸å˜)
// =====================================================================

std::vector<char> buildMsg(uint16_t msgid, const std::string &data)
{
    uint32_t len = data.size() + sizeof(uint16_t);
    std::vector<char> packet(sizeof(uint32_t) + len);

    uint32_t network_order_len = htonl(len);
    memcpy(packet.data(), &network_order_len, sizeof(uint32_t));

    uint16_t network_order_msgid = htons(msgid);
    memcpy(packet.data() + sizeof(uint32_t), &network_order_msgid, sizeof(uint16_t));

    memcpy(packet.data() + sizeof(uint32_t) + sizeof(uint16_t), data.data(), data.size());

    return packet;
}

// =====================================================================
// å¼‚æ­¥ I/O æ ¸å¿ƒå‡½æ•°
// =====================================================================

// å£°æ˜ï¼Œç”¨äºé€’å½’è°ƒç”¨
void start_async_read(tcp::socket &socket);

/**
 * @brief æ¶ˆæ¯è§£æä¸åˆ†å‘ (åœ¨ io_context çº¿ç¨‹ä¸­æ‰§è¡Œ)
 * * @param socket å½“å‰è¿æ¥çš„ Socket
 * @param buffer æ¥æ”¶åˆ°çš„å®Œæ•´æ¶ˆæ¯ä½“ (ä¸å« 4 å­—èŠ‚é•¿åº¦å¤´)
 */
void handle_message(tcp::socket &socket, const std::vector<char> &buffer)
{
    if (buffer.size() < sizeof(uint16_t))
    {
        std::cerr << "[Client ERROR] Received truncated message." << std::endl;
        return;
    }

    // è§£ææ¶ˆæ¯ ID
    uint16_t network_order_msgid = 0;
    memcpy(&network_order_msgid, buffer.data(), sizeof(uint16_t));
    uint16_t msgid = ntohs(network_order_msgid);

    // å‡è®¾ body æ˜¯é™¤å» MsgID çš„éƒ¨åˆ†
    std::string payload(buffer.data() + sizeof(uint16_t), buffer.size() - sizeof(uint16_t));

    bool is_response = false;
    uint32_t matched_req_id = 0;

    // 1. å°è¯•å°†æ¶ˆæ¯ä½œä¸º**è¯·æ±‚å“åº”**å¤„ç† (æ£€æŸ¥æ˜¯å¦æœ‰çº¿ç¨‹åœ¨ç­‰å¾…)
    // é”ä¿æŠ¤å…±äº«çŠ¶æ€ map
    {
        std::lock_guard<std::mutex> lock(g_waiters_mutex);

        // éå† mapï¼ŒæŸ¥æ‰¾æ˜¯å¦æœ‰çº¿ç¨‹åœ¨ç­‰å¾…è¿™ä¸ª MsgID çš„å“åº”
        // æ³¨æ„ï¼šç”Ÿäº§ç¯å¢ƒä¸­æœ€å¥½ä½¿ç”¨ Request ID æŸ¥æ‰¾
        for (auto it = g_response_waiters.begin(); it != g_response_waiters.end();)
        {
            if (msgid == it->second->expected_msgid)
            {
                // åŒ¹é…æˆåŠŸï¼è¿™æ˜¯ä¸€ä¸ªæˆ‘ä»¬æ­£åœ¨ç­‰å¾…çš„å“åº”
                matched_req_id = it->first;
                std::cout << "[Client DEBUG] Received expected Response (ReqID:" << matched_req_id << ", MsgID:" << msgid << ")" << std::endl;

                // è®¾ç½® promise çš„å€¼ï¼Œè¿™å°†è§£é™¤ send_request çº¿ç¨‹çš„é˜»å¡
                // æ³¨æ„ï¼šåœ¨ io_context çº¿ç¨‹ä¸­è°ƒç”¨ set_value æ˜¯çº¿ç¨‹å®‰å…¨çš„ã€‚
                it->second->promise.set_value(buffer);

                // ç§»é™¤å·²å¤„ç†çš„ç­‰å¾…è€…
                it = g_response_waiters.erase(it);
                is_response = true;
                break;
            }
            else
            {
                ++it;
            }
        }
    } // é”é‡Šæ”¾

    // 2. å¦‚æœä¸æ˜¯å“åº”ï¼Œåˆ™ä½œä¸ºæœåŠ¡å™¨æ¨é€å¤„ç†
    if (!is_response)
    {
        std::cout << "\n--- [æœåŠ¡å™¨æ¨é€: MSG_ID " << msgid << "] ---\n";

        switch (msgid)
        {
        case MSG_CHAT:
        {
            msg::SChatMsg chat;
            if (chat.ParseFromString(payload))
                std::cout << "[èŠå¤©] ç©å®¶ " << chat.from() << ": " << chat.text() << std::endl;
            break;
        }
        case MSG_ADDEXPACK:
        {
            // æ³¨æ„ï¼šå¦‚æœæœåŠ¡å™¨å°† ACK ä½œä¸ºæ¨é€å‘é€ï¼Œä¸”ä¸»çº¿ç¨‹æ²¡æœ‰ç­‰å¾…ï¼Œåˆ™åœ¨è¿™é‡Œå¤„ç†
            msg::AddExpRsp rsp;
            if (rsp.ParseFromString(payload))
                std::cout << "[ç»éªŒ] å½“å‰ç»éªŒ: " << rsp.new_exp()
                          << " ç­‰çº§: " << rsp.new_level()
                          << (rsp.level_up() ? " ğŸ‰å‡çº§!" : "") << std::endl;
            break;
        }
        case MSG_ENTER_ROOM_ACK:
        case MSG_READY_ACK:
        {
            // å¯¹äºéé¢„æœŸçš„ ACK æ¶ˆæ¯ï¼Œæˆ‘ä»¬åªæ‰“å°ï¼Œä¸å¤„ç†ï¼Œå› ä¸ºå®ƒä»¬å¯èƒ½è¢«ä¸»çº¿ç¨‹æ„å¤–å¿½ç•¥
            std::cout << "[æˆ¿é—´/å‡†å¤‡] æ”¶åˆ°æœªè¢«è¯·æ±‚çº¿ç¨‹æ¶ˆè€—çš„ ACK (" << msgid << ")." << std::endl;
            break;
        }
        case MSG_BATTLE_ACTION: // æˆ˜æ–—å¼€å§‹æ¨é€
        {
            msg::BattleStart bs;
            if (bs.ParseFromString(payload))
            {
                std::cout << "[æˆ˜æ–—å¼€å§‹] æˆ¿é—´ " << bs.roomid() << " ç©å®¶: ";
                for (int i = 0; i < bs.players_size(); ++i)
                    std::cout << bs.players(i) << " ";
                std::cout << std::endl;
            }
            break;
        }
        case MSG_BATTLE_SYNC:
        {
            msg::BattleSync sync;
            if (sync.ParseFromString(payload))
            {
                std::cout << "[æˆ˜æ–—åŒæ­¥] ";
                for (int i = 0; i < sync.states_size(); ++i)
                {
                    auto s = sync.states(i);
                    std::cout << "[uid=" << s.uid() << " HP=" << s.hp() << " MP=" << s.mp() << "] ";
                }
                std::cout << std::endl;
            }
            break;
        }
        default:
            std::cout << "[æœåŠ¡å™¨æ¨é€] æœªçŸ¥æ¶ˆæ¯ID=" << msgid << std::endl;
        }
        std::cout << "--- [æ¨é€å¤„ç†ç»“æŸ] ---\n";
    }
}

/**
 * @brief å¼‚æ­¥è¯»å–å¾ªç¯ (å–ä»£ receive_loop)
 */
void start_async_read(tcp::socket &socket)
{
    // ä½¿ç”¨ shared_ptr ç®¡ç† streambuf ç”Ÿå‘½å‘¨æœŸ
    auto buffer_ptr = std::make_shared<boost::asio::streambuf>();
    buffer_ptr->prepare(sizeof(uint32_t));

    // 1. å¼‚æ­¥è¯»å–é•¿åº¦ Header (4 å­—èŠ‚)
    boost::asio::async_read(socket, *buffer_ptr, boost::asio::transfer_exactly(sizeof(uint32_t)),
                            [&socket, buffer_ptr](boost::system::error_code ec, std::size_t length)
                            {
                                if (ec)
                                {
                                    if (ec != boost::asio::error::eof && ec != boost::asio::error::operation_aborted)
                                        std::cerr << "[Client ERROR] Header read failed: " << ec.message() << std::endl;
                                    return;
                                }

                                // æå–é•¿åº¦
                                uint32_t network_order_len;
                                const char *header_data = boost::asio::buffer_cast<const char *>(buffer_ptr->data());
                                memcpy(&network_order_len, header_data, sizeof(uint32_t));
                                uint32_t message_body_len = ntohl(network_order_len);

                                // ç®€å•çš„é•¿åº¦æ£€æŸ¥
                                if (message_body_len > 1024 * 1024 * 10 || message_body_len < sizeof(uint16_t))
                                {
                                    std::cerr << "[Client ERROR] Received abnormal message length: " << message_body_len << std::endl;
                                    return;
                                }

                                // 2. å¼‚æ­¥è¯»å– Body
                                // æ¶ˆè€—æ‰å·²è¯»å–çš„ 4 å­—èŠ‚ Header
                                buffer_ptr->consume(sizeof(uint32_t));
                                buffer_ptr->prepare(message_body_len); // å‡†å¤‡è¯»å– body çš„ç©ºé—´

                                boost::asio::async_read(socket, *buffer_ptr, boost::asio::transfer_exactly(message_body_len),
                                                        [&socket, buffer_ptr, message_body_len](boost::system::error_code ec_body, std::size_t length_body)
                                                        {
                                                            if (ec_body)
                                                            {
                                                                if (ec_body != boost::asio::error::eof && ec_body != boost::asio::error::operation_aborted)
                                                                    std::cerr << "[Client ERROR] Body read failed: " << ec_body.message() << std::endl;
                                                                return;
                                                            }

                                                            // 3. å®Œæ•´æ¶ˆæ¯æ¥æ”¶ï¼Œå¼€å§‹å¤„ç†
                                                            std::vector<char> full_message(message_body_len);
                                                            const char *body_data = boost::asio::buffer_cast<const char *>(buffer_ptr->data());
                                                            memcpy(full_message.data(), body_data, message_body_len);

                                                            // è°ƒç”¨æ¶ˆæ¯å¤„ç†å‡½æ•° (åœ¨ I/O çº¿ç¨‹ä¸­æ‰§è¡Œ)
                                                            handle_message(socket, full_message);

                                                            // 4. ç»§ç»­ä¸‹ä¸€ä¸ªå¼‚æ­¥è¯»å– (é€’å½’å¾ªç¯)
                                                            start_async_read(socket);
                                                        });
                            });
}

// ---------------------------------------------------------------------
// æ ¸å¿ƒä¿®å¤å‡½æ•°ï¼šsend_request (åŒæ­¥å‘é€è¯·æ±‚ï¼Œå¼‚æ­¥ç­‰å¾…å“åº”)
// ---------------------------------------------------------------------
std::vector<char> send_request(tcp::socket &socket, uint16_t msgid, const std::string &data)
{
    // 1. è®¾ç½®åŒæ­¥ç­‰å¾…æœºåˆ¶
    uint32_t current_req_id = ++g_request_counter;

    // åˆ›å»º ResponseWaiter å®ä¾‹
    auto waiter_ptr = std::make_shared<ResponseWaiter>();
    waiter_ptr->request_id = current_req_id;
    // å‡è®¾å“åº”IDæ˜¯ è¯·æ±‚ID + 1ï¼Œéœ€è¦æ ¹æ®æ‚¨çš„æœåŠ¡å™¨åè®®æ¥ç¡®å®š
    // å¦‚æœæœåŠ¡å™¨å“åº” ID å’Œè¯·æ±‚ ID ç›¸åŒï¼Œåˆ™è®¾ç½®ä¸º msgidã€‚
    // åœ¨è¿™é‡Œæˆ‘ä»¬å‡è®¾æœåŠ¡å™¨çš„ ACK ID æ˜¯å›ºå®šçš„ï¼Œä¾‹å¦‚ MSG_DENGLUACK å¯¹åº” MSG_DENGLU
    waiter_ptr->expected_msgid = msgid + 1;

    // è·å– futureï¼Œç”¨äºé˜»å¡ç­‰å¾…
    std::future<std::vector<char>> response_future = waiter_ptr->promise.get_future();

    // 2. å­˜å‚¨ç­‰å¾…è€…åˆ°å…±äº« Map ä¸­ (ä¿æŠ¤å…±äº«çŠ¶æ€)
    {
        std::lock_guard<std::mutex> lock(g_waiters_mutex);
        g_response_waiters[current_req_id] = waiter_ptr;
    }

    // 3. å‘é€è¯·æ±‚ - å†™å…¥æ“ä½œå¿…é¡»åŠ é”ï¼Œé˜²æ­¢ä¸å…¶å®ƒçº¿ç¨‹çš„å†™å…¥å†²çª (ä¾‹å¦‚å¦ä¸€ä¸ª send_request)
    std::lock_guard<std::mutex> lock(g_write_mutex);

    try
    {
        std::vector<char> package = buildMsg(msgid, data);
        std::cout << "[Client DEBUG] Sending request (ReqID:" << current_req_id << ", MsgID:" << msgid << "), size: " << package.size() << std::endl;

        // ä½¿ç”¨åŒæ­¥å†™å…¥ï¼Œç¡®ä¿è¯·æ±‚ä¸€æ¬¡æ€§å‘å‡º
        boost::asio::write(socket, boost::asio::buffer(package));
        std::cout << "[Client DEBUG] Request sent successfully. Waiting for async response." << std::endl;
    }
    catch (const std::exception &e)
    {
        std::cerr << "[Client ERROR] send_request write failed: " << e.what() << std::endl;

        // å†™å…¥å¤±è´¥ï¼Œå¿…é¡»æ¸…é™¤ç­‰å¾…è€…å¹¶è®¾ç½®å¼‚å¸¸
        {
            std::lock_guard<std::mutex> lock(g_waiters_mutex);
            auto it = g_response_waiters.find(current_req_id);
            if (it != g_response_waiters.end())
            {
                it->second->promise.set_exception(std::make_exception_ptr(std::runtime_error("Network write failed.")));
                g_response_waiters.erase(it);
            }
        }
        return std::vector<char>();
    }

    // 4. é˜»å¡ç­‰å¾…å¼‚æ­¥å›è°ƒè®¾ç½®ç»“æœ
    try
    {
        std::cout << "[Client DEBUG] Waiting for response for ReqID: " << current_req_id << std::endl;
        // future.get() é˜»å¡ä¸»çº¿ç¨‹ï¼Œç›´åˆ° I/O çº¿ç¨‹è°ƒç”¨ promise.set_value()
        std::vector<char> response = response_future.get();
        std::cout << "[Client DEBUG] Response received for ReqID: " << current_req_id << std::endl;
        return response;
    }
    catch (const std::exception &e)
    {
        // å¦‚æœ promise è®¾ç½®äº†å¼‚å¸¸ï¼Œæˆ– future å‡ºäº†é—®é¢˜
        std::cerr << "[Client ERROR] Synchonous wait failed: " << e.what() << std::endl;
        return std::vector<char>();
    }

    // æ³¨æ„ï¼šåŒ¹é…æˆåŠŸçš„ ResponseWaiter å·²ç»åœ¨ handle_message ä¸­è¢«ç§»é™¤äº†ã€‚
}

// ---------------------------------------------------------------------
// ä¸šåŠ¡é€»è¾‘å‡½æ•° (MsgID æ˜ å°„è°ƒæ•´ä¸ºå®é™… ID)
// ---------------------------------------------------------------------
// ... ä¿æŒä¸å˜ï¼Œä½†è¦ç¡®ä¿æ‰€æœ‰ send_request è°ƒç”¨çš„ MsgID ä¸æœåŠ¡å™¨ ACK ID åŒ¹é… ...
// ä¸ºäº†å…¼å®¹æ‚¨çš„åŸå§‹ä»£ç ï¼Œæˆ‘å°†ä¿æŒä¸šåŠ¡é€»è¾‘ä¸å˜ï¼Œå› ä¸ºå®ƒè°ƒç”¨äº† send_requestã€‚

bool do_register(tcp::socket &socket)
{
    msg::RegisterReq reg_req;
    std::string name, passwd;
    std::cout << "è¯·è¾“å…¥ç”¨æˆ·å: ";
    std::cin >> name;
    std::cout << "è¯·è¾“å…¥å¯†ç : ";
    std::cin >> passwd;
    reg_req.set_name(name);
    reg_req.set_passwd(passwd);
    std::string req_data;
    reg_req.SerializeToString(&req_data);

    // âš ï¸ ä¿®å¤ï¼šå°†æœŸæœ›çš„ ACK ID è®¾ç½®æ­£ç¡®
    // ç”±äº send_request ç°åœ¨å‡è®¾ ACK ID = REQ ID + 1ï¼Œä½†åœ¨ do_register ä¸­æˆ‘ä»¬çŸ¥é“ ACK ID æ˜¯ MSG_ZHUCEACK
    // ä¸ºäº†å…¼å®¹ï¼Œæˆ‘ä»¬åœ¨ do_register ä¸­åˆ›å»ºå¹¶å­˜å‚¨ç­‰å¾…è€…ï¼Œç›´æ¥è°ƒç”¨ writeï¼Œç„¶åç­‰å¾…

    std::vector<char> resp_buffer = send_request(socket, MSG_ZHUCE, req_data);
    if (resp_buffer.empty())
        return false;

    uint16_t network_order_msgid = 0;
    memcpy(&network_order_msgid, resp_buffer.data(), sizeof(uint16_t));
    uint16_t msgid = ntohs(network_order_msgid);

    // å‡è®¾æœåŠ¡å™¨è¿”å›çš„ ACK ID æ˜¯ MSG_ZHUCEACK
    if (msgid != MSG_ZHUCEACK)
    {
        std::cerr << "æ³¨å†Œå¤±è´¥ï¼Œæ”¶åˆ°éé¢„æœŸæ¶ˆæ¯ID: " << msgid << std::endl;
        return false;
    }
    std::string resp_data(resp_buffer.data() + sizeof(uint16_t), resp_buffer.size() - sizeof(uint16_t));
    msg::RegisterResp reg_resp;
    if (!reg_resp.ParseFromString(resp_data))
    {
        std::cerr << "æ³¨å†Œå“åº”è§£æå¤±è´¥" << std::endl;
        return false;
    }
    if (reg_resp.ok())
    {
        std::cout << "æ³¨å†ŒæˆåŠŸ! åˆ†é… UID: " << reg_resp.uid() << std::endl;
        return true;
    }
    else
    {
        std::cout << "æ³¨å†Œå¤±è´¥: " << reg_resp.reason() << std::endl;
        return false;
    }
}
bool do_login(tcp::socket &socket)
{
    msg::LoginReq login_req;
    int uid;
    std::string passwd;
    std::cout << "è¯·è¾“å…¥ UID: ";
    std::cin >> uid;
    std::cout << "è¯·è¾“å…¥å¯†ç : ";
    std::cin >> passwd;
    login_req.set_uid(uid);
    login_req.set_passwd(passwd);
    std::string req_data;
    login_req.SerializeToString(&req_data);
    std::cout << "ç­‰å¾…æ¥å—æœåŠ¡å™¨å“åº”" << std::endl;

    // âš ï¸ ä¿®å¤ï¼šå°†æœŸæœ›çš„ ACK ID è®¾ç½®æ­£ç¡®
    std::vector<char> resp_buffer = send_request(socket, MSG_DENGLU, req_data);
    if (resp_buffer.empty())
        return false;

    uint16_t network_order_msgid = 0;
    memcpy(&network_order_msgid, resp_buffer.data(), sizeof(uint16_t));
    uint16_t msgid = ntohs(network_order_msgid);

    // å‡è®¾æœåŠ¡å™¨è¿”å›çš„ ACK ID æ˜¯ MSG_DENGLUACK
    if (msgid != MSG_DENGLUACK)
    {
        std::cerr << "ç™»å½•å¤±è´¥ï¼Œæ”¶åˆ°éé¢„æœŸæ¶ˆæ¯ID: " << msgid << std::endl;
        return false;
    }
    std::string resp_data(resp_buffer.data() + sizeof(uint16_t), resp_buffer.size() - sizeof(uint16_t));
    msg::LoginResp login_resp;
    if (!login_resp.ParseFromString(resp_data))
    {
        std::cerr << "ç™»å½•å“åº”è§£æå¤±è´¥" << std::endl;
        return false;
    }
    if (login_resp.ok())
    {
        Currentid_ = uid;
        std::cout << "ç™»å½•æˆåŠŸ!" << std::endl;
        return true;
    }
    else
    {
        std::cout << "ç™»å½•å¤±è´¥: " << login_resp.reason() << std::endl;
        return false;
    }
}
bool do_view_data(tcp::socket &socket)
{
    msg::ViewPlayerDataReq req;
    req.set_uid(Currentid_);

    std::string req_data;
    req.SerializeToString(&req_data);

    std::vector<char> resp_buffer = send_request(socket, MSG_BACKPACK, req_data);
    if (resp_buffer.empty())
        return false;

    uint16_t network_order_msgid = 0;
    memcpy(&network_order_msgid, resp_buffer.data(), sizeof(uint16_t));
    uint16_t msgid = ntohs(network_order_msgid);

    if (msgid != MSG_BACKPACKACK)
    {
        std::cerr << "æŸ¥çœ‹ç©å®¶æ•°æ®å¤±è´¥ï¼Œæ”¶åˆ°éé¢„æœŸæ¶ˆæ¯ID: " << msgid << std::endl;
        return false;
    }

    std::string resp_data(resp_buffer.data() + sizeof(uint16_t), resp_buffer.size() - sizeof(uint16_t));
    msg::PlayerAttr player_data;
    if (!player_data.ParseFromString(resp_data))
    {
        std::cerr << "æŸ¥çœ‹ç©å®¶æ•°æ®å“åº”è§£æå¤±è´¥" << std::endl;
        return false;
    }

    std::cout << "ğŸ¹ ç©å®¶æ•°æ®:\n";
    std::cout << " Â UID: " << player_data.uid() << "\n";
    std::cout << " Â ç­‰çº§: " << player_data.level() << "\n";
    std::cout << " Â ç»éªŒ: " << player_data.exp() << "\n";
    std::cout << " Â HP/MP: " << player_data.hp() << "/" << player_data.mp() << "\n";
    std::cout << " Â é‡‘å¸: " << player_data.coin() << "\n";
    std::cout << " Â åæ ‡: (" << player_data.x() << ", " << player_data.y() << ", " << player_data.z() << ")\n";
    return true;
}
bool do_add_exp(tcp::socket &socket)
{
    int add_value;
    std::cout << "è¯·è¾“å…¥å¢åŠ çš„ç»éªŒå€¼: ";
    std::cin >> add_value;

    msg::AddExpReq req;
    req.set_uid(Currentid_);
    req.set_exp_add(add_value);

    std::string req_data;
    req.SerializeToString(&req_data);

    std::vector<char> resp_buffer = send_request(socket, MSG_ADDEXP, req_data);
    if (resp_buffer.empty())
        return false;

    uint16_t network_order_msgid = 0;
    memcpy(&network_order_msgid, resp_buffer.data(), sizeof(uint16_t));
    uint16_t msgid = ntohs(network_order_msgid);

    if (msgid != MSG_ADDEXPACK)
    {
        std::cerr << "å¢åŠ ç»éªŒå¤±è´¥ï¼Œæ”¶åˆ°éé¢„æœŸæ¶ˆæ¯ID: " << msgid << std::endl;
        return false;
    }

    msg::AddExpRsp rsp;
    std::string resp_data(resp_buffer.data() + sizeof(uint16_t), resp_buffer.size() - sizeof(uint16_t));
    if (!rsp.ParseFromString(resp_data))
    {
        std::cerr << "AddExp å“åº”è§£æå¤±è´¥" << std::endl;
        return false;
    }

    if (rsp.success())
    {
        std::cout << "âœ… å¢åŠ ç»éªŒæˆåŠŸ! å½“å‰ç»éªŒ: " << rsp.new_exp()
                  << ", ç­‰çº§: " << rsp.new_level()
                  << (rsp.level_up() ? " ğŸ‰å‡çº§å•¦!" : "") << std::endl;
        return true;
    }
    else
    {
        std::cerr << "âŒ å¢åŠ ç»éªŒå¤±è´¥!" << std::endl;
        return false;
    }
}
void enterRoom(tcp::socket &socket, int roomid)
{
    msg::EnterRoomReq req;
    req.set_uid(Currentid_);
    req.set_roomid(roomid);

    std::vector<char> resp_buffer = send_request(socket, MSG_ENTER_ROOM, req.SerializeAsString());
    if (resp_buffer.empty())
    {
        std::cerr << "âŒ è¿›å…¥æˆ¿é—´è¯·æ±‚å¤±è´¥æˆ–è¿æ¥æ–­å¼€ã€‚" << std::endl;
        return;
    }

    // Parse response
    uint16_t network_order_msgid = 0;
    memcpy(&network_order_msgid, resp_buffer.data(), sizeof(uint16_t));
    uint16_t msgid = ntohs(network_order_msgid);

    if (msgid != MSG_ENTER_ROOM_ACK)
    {
        std::cerr << "âŒ è¿›å…¥æˆ¿é—´å¤±è´¥ï¼Œæ”¶åˆ°éé¢„æœŸæ¶ˆæ¯ID: " << msgid << std::endl;
        return;
    }

    std::string resp_data(resp_buffer.data() + sizeof(uint16_t), resp_buffer.size() - sizeof(uint16_t));
    msg::EnterRoomAck ack; // å‡è®¾å“åº”æ¶ˆæ¯ä¸º EnterRoomAck
    if (!ack.ParseFromString(resp_data))
    {
        std::cerr << "âŒ è¿›å…¥æˆ¿é—´å“åº”è§£æå¤±è´¥ã€‚" << std::endl;
        return;
    }

    if (ack.ok())
    {
        std::cout << "âœ… æˆåŠŸè¿›å…¥æˆ¿é—´ " << roomid;
        std::cout << std::endl;
    }
    else
    {
        std::cerr << "âŒ è¿›å…¥æˆ¿é—´ " << roomid << " å¤±è´¥: " << ack.reason() << std::endl;
    }
}
void ready(tcp::socket &socket, int roomid, bool is_ready)
{
    msg::ReadyReq req;
    req.set_uid(Currentid_);
    req.set_roomid(roomid);
    req.set_ready(is_ready);

    std::vector<char> resp_buffer = send_request(socket, MSG_READY, req.SerializeAsString());
    if (resp_buffer.empty())
    {
        std::cerr << "âŒ å‡†å¤‡è¯·æ±‚å¤±è´¥æˆ–è¿æ¥æ–­å¼€ã€‚" << std::endl;
        return;
    }

    uint16_t network_order_msgid = 0;
    memcpy(&network_order_msgid, resp_buffer.data(), sizeof(uint16_t));
    uint16_t msgid = ntohs(network_order_msgid);

    if (msgid != MSG_READY_ACK)
    {
        std::cerr << "âŒ å‡†å¤‡åŠ¨ä½œå¤±è´¥ï¼Œæ”¶åˆ°éé¢„æœŸæ¶ˆæ¯ID: " << msgid << std::endl;
        return;
    }

    std::string resp_data(resp_buffer.data() + sizeof(uint16_t), resp_buffer.size() - sizeof(uint16_t));
    msg::ReadyAck ack; // å‡è®¾å“åº”æ¶ˆæ¯ä¸º ReadyAck
    if (!ack.ParseFromString(resp_data))
    {
        std::cerr << "âŒ å‡†å¤‡å“åº”è§£æå¤±è´¥ã€‚" << std::endl;
        return;
    }

    std::cout << "âœ… æˆ¿é—´ " << roomid << " å‡†å¤‡åŠ¨ä½œæˆåŠŸ: "
              << (is_ready ? "å·²å‡†å¤‡" : "å–æ¶ˆå‡†å¤‡") << std::endl;
}

void battleAction(tcp::socket &socket, int roomid, int skillid, int target)
{
    // å‡è®¾å®¢æˆ·ç«¯å‘é€çš„è¯·æ±‚ç»“æ„ä½“æ˜¯ msg::BattleActionReq
    msg::BattleAction req;
    req.set_uid(Currentid_);
    req.set_roomid(roomid);
    req.set_skillid(skillid);
    req.set_target(target);

    std::string req_data;
    req.SerializeToString(&req_data);

    // æˆ˜æ–—åŠ¨ä½œé€šå¸¸æ˜¯ Fire-and-Forgetï¼Œç»“æœé€šè¿‡ MSG_BATTLE_SYNC å¼‚æ­¥æ¨é€ã€‚
    // æˆ‘ä»¬åªåé¦ˆå‘é€æ˜¯å¦æˆåŠŸã€‚
    std::lock_guard<std::mutex> lock(g_write_mutex);

    try
    {
        std::vector<char> package = buildMsg(MSG_BATTLE_ACTION, req_data);
        boost::asio::write(socket, boost::asio::buffer(package));
        std::cout << "âœ… æˆ˜æ–—åŠ¨ä½œ (SkillID: " << skillid << ") å‘é€æˆåŠŸ! ç­‰å¾…æœåŠ¡å™¨åŒæ­¥ç»“æœ..." << std::endl;
    }
    catch (const std::exception &e)
    {
        std::cerr << "âŒ æˆ˜æ–—åŠ¨ä½œå‘é€å¤±è´¥: " << e.what() << std::endl;
    }
}

// ---------------------------------------------------------------------
// ä¸»å‡½æ•°
// ---------------------------------------------------------------------
int main(int argc, char *argv[])
{
    if (argc != 3)
    {
        std::cerr << "ç”¨æ³•: " << argv[0] << " <host> <port>\n";
        return 1;
    }

    // å¿…é¡»åœ¨ try å—å¤–éƒ¨å®šä¹‰ io_context å’Œ socket
    boost::asio::io_context io;
    tcp::socket socket(io);

    try
    {
        tcp::resolver resolver(io);
        boost::asio::connect(socket, resolver.resolve(argv[1], argv[2]));
        std::cout << "âœ… è¿æ¥æœåŠ¡å™¨æˆåŠŸ\n";

        // 1. å¯åŠ¨ I/O çº¿ç¨‹æ± 
        boost::asio::executor_work_guard<boost::asio::io_context::executor_type> work_guard(io.get_executor());

        std::vector<std::thread> io_threads;
        int num_threads = std::thread::hardware_concurrency() > 0 ? std::thread::hardware_concurrency() : 1;
        for (int i = 0; i < num_threads; ++i)
        {
            io_threads.emplace_back([&io]()
                                    { 
                try {
                    io.run(); 
                } catch (const std::exception& e) {
                    std::cerr << "[I/O Thread ERROR] " << e.what() << std::endl;
                } });
        }
        std::cout << "[Client DEBUG] Started " << num_threads << " I/O threads." << std::endl;

        // 2. å¯åŠ¨ç¬¬ä¸€ä¸ªå¼‚æ­¥è¯»å–æ“ä½œ
        start_async_read(socket);

        // ä¸»å¾ªç¯ (ä¸šåŠ¡é€»è¾‘)
        while (true)
        {
            int choice;
            std::cout << "\næ“ä½œé€‰æ‹©: \n1-æ³¨å†Œ 2-ç™»å½• 3-æŸ¥çœ‹æ•°æ® 4-å¢åŠ ç»éªŒ 5-è¿›å…¥æˆ¿é—´ 6-å‡†å¤‡ 7-æˆ˜æ–—åŠ¨ä½œ 0-é€€å‡º\n> ";
            if (!(std::cin >> choice))
            {
                std::cout << "è¾“å…¥é”™è¯¯ï¼Œé€€å‡ºç¨‹åºã€‚\n";
                break;
            }

            if (choice == 1)
                do_register(socket);
            else if (choice == 2)
                do_login(socket);
            else if (choice == 3)
            {
                if (Currentid_ == -1)
                {
                    std::cout << "è¯·å…ˆç™»å½•ã€‚\n";
                    continue;
                }
                do_view_data(socket);
            }
            else if (choice == 4)
            {
                if (Currentid_ == -1)
                {
                    std::cout << "è¯·å…ˆç™»å½•ã€‚\n";
                    continue;
                }
                do_add_exp(socket);
            }
            else if (choice == 5)
            {
                if (Currentid_ == -1)
                {
                    std::cout << "è¯·å…ˆç™»å½•ã€‚\n";
                    continue;
                }
                int roomid;
                std::cout << "è¾“å…¥æˆ¿é—´ID: ";
                std::cin >> roomid;
                enterRoom(socket, roomid);
            }
            else if (choice == 6)
            {
                if (Currentid_ == -1)
                {
                    std::cout << "è¯·å…ˆç™»å½•ã€‚\n";
                    continue;
                }
                int roomid;
                std::cout << "è¾“å…¥æˆ¿é—´ID: ";
                std::cin >> roomid;
                ready(socket, roomid, true);
            }
            else if (choice == 7)
            {
                if (Currentid_ == -1)
                {
                    std::cout << "è¯·å…ˆç™»å½•ã€‚\n";
                    continue;
                }
                int roomid, skillid, target;
                std::cout << "è¾“å…¥æˆ¿é—´ID æŠ€èƒ½ID ç›®æ ‡UID: ";
                std::cin >> roomid >> skillid >> target;
                battleAction(socket, roomid, skillid, target);
            }
            else if (choice == 0)
                break;
        }

        // æ¸…ç†èµ„æº
        socket.close();
        io.stop();
        for (auto &t : io_threads)
        {
            if (t.joinable())
                t.join();
        }
    }
    catch (const std::exception &e)
    {
        std::cerr << "å®¢æˆ·ç«¯å¼‚å¸¸: " << e.what() << std::endl;
    }
    return 0;
}